File: medium _high/club/club.cpp
#include <iostream>
#include <vector>
#include <queue>
using ll=long long;
using namespace std;
ll n,m;
vector<vector<ll>> ship;
int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cin>>n>>m;
    ship.resize(n);
    vector<ll> cnt(n,0LL);
    for (ll i=0;i<m;i++)
    {
        ll a,b; cin>>a>>b; a--; b--;
        ship[a].push_back(b);
        cnt[b]++;
    }
    queue<ll> temp;
    for (ll i=0;i<n;i++)
        if (!cnt[i])
            temp.push(i);
    vector<ll> path;
    while (!temp.empty())
    {
        ll pwd=temp.front();
        temp.pop();
        path.push_back(pwd);
        for (ll chd:ship[pwd])
        {
            cnt[chd]--;
            if (!cnt[chd])
                temp.push(chd);
        }
    }
    if ((ll)path.size()!=n) cout<<"NO";
    else
    {
        cout<<"YES"<<"\n";
        for (ll i=0;i<n;i++)
        {
            cout<<path[i]+1LL;
            if (i!=n-1LL)
                cout<<"\n";
        }
    }
    return 0;
}

File: medium _high/club/club.py
import sys
from collections import deque
input=sys.stdin.readline
n,m=map(int,input().split())
ship=[[] for _ in range(n)]
temp=deque()
cnt=[0 for _ in range(n)]
for _ in range(m):
	u,v=map(int,input().split())
	ship[u-1].append(v-1)
	cnt[v-1]+=1
for i in range(n):
	if cnt[i]==0:
		temp.append(i)
path=[]
while temp:
	pwd=temp.popleft()
	path.append(pwd+1)
	for nxt in ship[pwd]:
		cnt[nxt]-=1
		if cnt[nxt]==0:
			temp.append(nxt)
if len(path)!=n:
	print("NO")
else:
	print("YES")
	ans="\n".join(map(str,path))
	print(ans)